src/libraries/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -13,9 +13,7 @@ public static partial class Enumerable
         static partial void CreateSelectIPartitionIterator<TResult, TSource>(
             Func<TSource, TResult> selector, IPartition<TSource> partition, ref IEnumerable<TResult>? result)
         {
-            result = partition is EmptyPartition<TSource> ?
-                EmptyPartition<TResult>.Instance :
-                new SelectIPartitionIterator<TSource, TResult>(partition, selector);
+            result = new SelectIPartitionIterator<TSource, TResult>(partition, selector);
```

Comment:
This seems like an allocation regression if not a performance regression. In the old code we would directly return an empty iterator for the `TResult` if the source was an empty iterator for a `TSource`. In the new code, we will actually allocate the partitioning iterator, which will then expose the inner emptiness.

Reply:
I was looking only at the changeset, sorry. 

Since this seems to be called [exactly one place](https://github.com/dotnet/runtime/pull/88425/files#diff-c94d44124e0ba957ce662c15b67da70da783cfc029dd0aa30e3b6427b6407cd2R56) now, and is a trivial one-liner, would it be clearer to just inline it there? Seems like this would be much simpler to read at the call site.

In other words, doesn't the code block in [Linq.Select.cs](https://github.com/dotnet/runtime/pull/88425/files#diff-c94d44124e0ba957ce662c15b67da70da783cfc029dd0aa30e3b6427b6407cd2R53-R61) now reduce to:

```csharp
if (source is IPartition<TSource> partition)
{
    return new SelectIPartitionIterator<TSource, TResult>(partition, selector);
}
```

I am not fully grokking the motivation for this `.SpeedOpt` file, but seems to me that as written currently, this code cannot return a null and a find-in-files on the whole runtime source is not surfacing another similar function.

Reply:
> I am not fully grokking the motivation for this .SpeedOpt file

There are multiple builds, one of which builds in additional optimizations that take up non-trivial size and thus one that prefers size over speed.

> would it be clearer to just inline it there?

It can't be without breaking the build.

Reply:
Cool, that explains why it's in SpeedOpt. Thanks

I see no other version or calls to that method, am I understanding that if the .SpeedOpt version isn't compiled & linked in, that's just a NoOp call. Sorry, still learning the new partial fill-ins...

Reply:
> am I understanding that if the .SpeedOpt version isn't compiled & linked in, that's just a NoOp call

[Yes.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BiAOwFcAbS7YSmAAgBMBLXW+gWACgABAJgCMPXgGYGAB2xQMLbJQYCGAYQYBvHgy2LxLUhgYBZABQBKdZu1WA8sABWMMAYgMAvAwrUA3JataAKgAWbACSuABKMPgQAG4wTMawAGYMEKY+3H7aLCnGLmweVJTmvADsDIIZWVplDAAMVdoAvr4MrVIycgq8KAxBoRFRsfGJMCkQ9o7O6TxNQA===)

Reply:
Cool thanks for teaching me!

Sorry, as a last question, as a newbie to the syntax, would it also have been legal to have the default implementation of the `CreateSelectIPartitionIterator` just return a `new SelectEnumerableIterator<TSource, TResult>(source, selector)` since that's what it is going to eventually fall through to? I know that would prevent other optimizations being added _later_  to the Linq\Select method and that it would prevent the compiler from discarding the type-cast and test, so we wouldn't ACTUALLY want to do that here, just asking about it conceptually.

Thanks again for the education!

Reply:
Depends on what you mean by "default implementation".  You can only have a single implementation of a partial method in the compilation unit.  If we wanted to define the method in a size-optimized-specific-file as well, then sure, each method could do something specific to that implementation.  It also wouldn't need to be partial at that point.